test-app/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt

```diff
@@ -0,0 +1,409 @@
+            } else {
+                maneuverView.render(ManeuverState.ManeuverSub.Hide)
+            }
+            maneuverView.render(ManeuverState.ManeuverPrimary.Instruction(primary))
+            maneuverView.render(ManeuverState.UpcomingManeuvers.RemoveUpcoming(maneuver))
+            if (lane != null) {
+                maneuverView.render(ManeuverState.LaneGuidanceManeuver.Show)
+                maneuverView.render(ManeuverState.LaneGuidanceManeuver.AddLanes(lane))
+            } else {
+                maneuverView.render(ManeuverState.LaneGuidanceManeuver.Hide)
+                maneuverView.render(ManeuverState.LaneGuidanceManeuver.RemoveLanes)
+            }
+        }
+    }
+
+    private val stepDistanceRemainingCallback = object : StepDistanceRemainingCallback {
+        override fun onStepDistanceRemaining(
+            distanceRemaining: ManeuverState.DistanceRemainingToFinishStep
+        ) {
+            maneuverView.render(distanceRemaining)
+        }
+    }
+
+    private val upcomingManeuversCallback = object : UpcomingManeuversCallback {
+        override fun onUpcomingManeuvers(state: ManeuverState.UpcomingManeuvers.Upcoming) {
+            maneuverView.render(state)
+        }
+    }
+
+    private val locationObserver = object : LocationObserver {
+        override fun onRawLocationChanged(rawLocation: Location) { }
+        override fun onEnhancedLocationChanged(
+            enhancedLocation: Location,
+            keyPoints: List<Location>
+        ) {
+            navigationLocationProvider.changePosition(
+                enhancedLocation,
+                keyPoints,
+            )
+            viewportDataSource.onLocationChanged(enhancedLocation)
+            viewportDataSource.evaluate()
+            navigationCamera.requestNavigationCameraToFollowing()
+        }
+    }
+
+    private val routesObserver = object : RoutesObserver {
+        override fun onRoutesChanged(routes: List<DirectionsRoute>) {
+            if (routes.isNotEmpty()) {
+                start.visibility = VISIBLE
+                directionsRoute = routes[0]
+                updateCameraToOverview()
+                val routeLeg = routes.first().legs()
+                ifNonNull(routeLeg) {
+                    maneuverApi.retrieveUpcomingManeuvers(it.first(), upcomingManeuversCallback)
+                }
+            } else {
+                updateCameraToIdle()
+            }
+        }
+    }
+
+    private val routeProgressObserver = object : RouteProgressObserver {
+        override fun onRouteProgressChanged(routeProgress: RouteProgress) {
+            viewportDataSource.onRouteProgressChanged(routeProgress)
+            viewportDataSource.evaluate()
+            tripProgressView.render(tripProgressApi.getTripProgress(routeProgress))
+            ifNonNull(routeProgress.currentLegProgress) { legProgress ->
+                ifNonNull(legProgress.currentStepProgress) {
```

Comment:
Maybe room for improvement for the maneuver API to do this automatically?

Reply:
This was intentionally done or else the API has to manage sending appropriate callbacks with messages.